File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        mem = [-1] * len(nums)
        def dfs(i) -> int:
            if i >= len(nums):
                return 0
            if mem[i] < 0:
                mem[i] = max(nums[i] + dfs(i + 2), dfs(i + 1))
            return mem[i]
        return dfs(0)

File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {

    private void dfs(int[] nums, int target, List<List<Integer>> resultList, int i, List<Integer> currentList, int total) {
        if (total == target) {
            resultList.add(new ArrayList<>(currentList));
            return;
        }
        if (i >= nums.length || total > target) {
            return;
        }
        dfs(nums, target, resultList, i + 1, currentList, total);
        currentList.add(nums[i]);
        dfs(nums, target, resultList, i, currentList, total + nums[i]);
        currentList.removeLast();
    }

    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> resultList = new ArrayList<>();
        dfs(nums, target, resultList, 0, new ArrayList<Integer>(), 0);
        return resultList;
    }
}